internal/objects/reconciler.go

```diff
@@ -108,21 +104,50 @@ type HNCConfigReconcilerType interface {
 //
 // +kubebuilder:rbac:groups=*,resources=*,verbs=*
 
-// SyncNamespace can be called manually by the HierarchyConfigReconciler when the hierarchy changes.
-// It enqueues all the current objects in the namespace and local copies of the original objects
-// in the ancestors.
-func (r *Reconciler) SyncNamespace(ctx context.Context, log logr.Logger, ns string) error {
+// OnChangeNamespace is called whenever the a namespace is changed (according to
+// HierarchyConfigReconciler). It enqueues all the current objects in the namespace and local copies
+// of the original objects in the ancestors.
+func (r *Reconciler) OnChangeNamespace(log logr.Logger, ns *forest.Namespace) {
 	log = log.WithValues("gvk", r.GVK)
+	nsnm := ns.Name()
+	srcnms := ns.Parent().GetAncestorSourceNames(r.GVK, "")
 
 	// Enqueue all the current objects in the namespace because some of them may have been deleted.
-	if err := r.enqueueLocalObjects(ctx, log, ns); err != nil {
-		return err
-	}
+	go func() {
+		// Usually, if we hit an error, we just return it and let controller-runtime worry about it. But
+		// since this is in a goroutine, we can't return it. Therefore, we need to handle the error
+		// ourselves.
+		attempts := 0
+		for {
+			attempts++
+			err := r.enqueueExistingObjects(context.Background(), log, nsnm)
+			if err == nil {
+				break
+			}
+			log.Error(err, "Couldn't list objects")
+			time.Sleep(10 * time.Second)
+			if attempts == 3 {
+				// We can get into a permanent failure loop if, for example, the GVK has been deleted from
+				// the server (i.e. it was a CRD). Rather than looping forever, let's restart and hope that
+				// fixes it.
+				log.Info("Couldn't list objects for 30s. Something's gone very wrong; hopefully a restart will fix it.")
```

Comment:
My concern here is that when we hit this, HNC will restart, which means the webhooks wont respond. If the api server is under heavy load (which I've seen a lot in the past on new cluster bring ups), it could lead us into weird loop of HNC being down, so webhooks fail, so things don't clear from api server queues, so HNC has issues listing objects. 

I don't think it is a huge concern, but I have had to temporarily delete webhooks for HNC, Kyverno, OPA, etc in the past due to some of these issues to get things to recover, which is never ideal. Do you think it's worthwhile to expose MaxRetryTimeout as a cli flag so it can be adjusted?

Reply:
Yeah maybe I should just take this out and see if it ever becomes a problem? E.g. the goroutine should still terminate but maybe give any _one_ goroutine a 60s timeout and then have it just exit.

We should probably look more holistically at things that would cause permanent failures in the object reconcilers, e.g. having a CRD deleted, and see how best to address that.